### PR TITLE
Bjorncs/system flags handler

### PR DESCRIPTION
**Pull request by bjorncs**
Initial version of the /system-flags/v1 api. Access control, bindings and proper error handling will come in future PR.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/systemflags/FlagsClient.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        try {
+            return new URIBuilder(target.endpoint()).setPath(FLAGS_V1_PATH + subPath).build();
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e); // should never happen
+        }
+    }
+
+    private static void verifySuccess(HttpResponse response, FlagsTarget target, FlagId flagId) throws IOException {
+        if (!success(response)) {
+            String message = getErrorMessage(response);
```
Comment (hakonhall) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/systemflags/FlagsClient.java]: should extract the error-code too?

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/ConfigserverFlagsTarget.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import static com.yahoo.vespa.hosted.controller.api.systemflags.v1.FlagsTarget.defaultFile;
+import static com.yahoo.vespa.hosted.controller.api.systemflags.v1.FlagsTarget.environmentFile;
+import static com.yahoo.vespa.hosted.controller.api.systemflags.v1.FlagsTarget.systemFile;
+import static com.yahoo.vespa.hosted.controller.api.systemflags.v1.FlagsTarget.zoneFile;
+
+/**
+ * @author bjorncs
+ */
+class ConfigserverFlagsTarget implements FlagsTarget {
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/ConfigserverFlagsTarget.java]: Fixed

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/SystemFlagsDataArchive.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+import static com.yahoo.yolean.Exceptions.uncheck;
+
+/**
+ * Represents a hierarchy of flag data files. See {@link FlagsTarget} for file naming convention.
+ *
+ * The flag files must reside in a 'flags/' root directory containing a directory for each flag name:
+ * {@code ./flags/<feature-flag-name>/*.json}
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/SystemFlagsDataArchive.java]: Fixed

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/systemflags/FlagsClient.java (hunk 1 above)
Comment (bjorncs) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/systemflags/FlagsClient.java]: Fixed

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/wire/SystemFlagsV1Api.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * @author bjorncs
+ */
+@Path("/system-flags/v1")
+public interface SystemFlagsV1Api {
+
+    @PUT
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes("application/zip")
+    @Path("/deploy")
+    WireSystemFlagsDeployResult deploy(@QueryParam("dryRun") @DefaultValue("false") boolean dryRun, InputStream inputStream);
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/wire/SystemFlagsV1Api.java]: Completed

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/wire/SystemFlagsV1Api.java [hunk 5]
```diff
@@ -20,6 +18,12 @@
...
     @Consumes("application/zip")
     @Path("/deploy")
-    WireSystemFlagsDeployResult deploy(@QueryParam("dryRun") @DefaultValue("false") boolean dryRun, InputStream inputStream);
+    WireSystemFlagsDeployResult deploy(InputStream inputStream);
+
+    @PUT
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes("application/zip")
+    @Path("/dryrun")
+    WireSystemFlagsDeployResult dryrun(InputStream inputStream);
```
Comment (hakonhall) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/systemflags/v1/wire/SystemFlagsV1Api.java]: "dryrun" is not very descriptive compared to "deploy", especially if /system-flags/v1 gets more path bindings. Should probably be called "deployprobe" or something?

**Review by hakonhall — APPROVED** [review]

**Pull request merged**